coil-fakeimageloader/src/main/java/coil/FakeImageLoader.kt

```diff
@@ -0,0 +1,232 @@
+    coroutineContext: CoroutineContext,
+    override val defaults: DefaultRequestOptions,
+    override val components: ComponentRegistry,
+    override val memoryCache: MemoryCache,
+    override val diskCache: DiskCache,
+    val engine: Engine,
+) : ImageLoader {
+
+    private val scope = CoroutineScope(coroutineContext)
+    private val shutdown = AtomicBoolean(false)
+
+    private val _requests = MutableSharedFlow<ImageRequest>()
+    val requests: Flow<ImageRequest> get() = _requests
+
+    private val _results = MutableSharedFlow<ImageResult>()
+    val results: Flow<ImageResult> get() = _results
+
+    override fun enqueue(request: ImageRequest): Disposable {
+        startCommon(request)
+        return JobDisposable(job = scope.async { executeCommon(request) })
+    }
+
+    override suspend fun execute(request: ImageRequest): ImageResult {
+        startCommon(request)
+        return executeCommon(request)
+    }
+
+    private fun startCommon(request: ImageRequest) {
+        assertNotShutdown()
+        assertValidRequest(request)
+        _requests.tryEmit(request)
+    }
+
+    private suspend fun executeCommon(request: ImageRequest): ImageResult {
+        request.target?.onStart(request.placeholder)
+        val result = engine.execute(request)
+        _results.tryEmit(result)
+        return result
+    }
+
+    override fun shutdown() {
+        shutdown.set(true)
+    }
+
+    override fun newBuilder(): ImageLoader.Builder {
+        throw UnsupportedOperationException("newBuilder() is not supported")
+    }
+
+    fun assertNotShutdown() {
+        check(!shutdown.get()) { "image loader is shutdown" }
+    }
+
+    private fun assertValidRequest(request: ImageRequest) {
+        if (request.data === NullRequestData) {
+            throw NullRequestDataException()
+        }
+    }
+
+    fun interface Engine {
+        suspend fun execute(request: ImageRequest): ImageResult
+    }
+
+    class CompositeQueueEngine : Engine {
+
+        private val results = mutableMapOf<Any, Queue<Engine>>()
+        private var default: Engine? = null
+
+        @Synchronized
+        fun enqueue(
+            data: Any,
+            drawable: Drawable,
+            dataSource: DataSource = DataSource.DISK,
+            memoryCacheKey: MemoryCache.Key? = null,
+            diskCacheKey: String? = null,
+        ) = enqueue(
+            data = data,
+            engine = { request ->
+                SuccessResult(
+                    drawable = drawable,
+                    request = request,
+                    dataSource = dataSource,
+                    memoryCacheKey = memoryCacheKey,
+                    diskCacheKey = diskCacheKey
+                )
+            }
+        )
+
+        @Synchronized
+        fun enqueue(data: Any, engine: Engine) = apply {
+            results.getOrPut(data) { LinkedList() }.offer(engine)
+        }
+
+        @Synchronized
+        fun default(default: Engine) = apply {
```

Comment:
This design with Engine, is beautiful and probably deserves to be called out in the PR.

Seems great for implementing a FakeImageLoader in a real app that uses local images for screenshot testing without predicting all requests up front.

Reply:
Thanks! 🙏